pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,19 +215,97 @@ func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) [
 	return tasks
 }
 
-// SuccessfulPipelineTaskNames returns a list of the names of all of the PipelineTasks in state
-// which have successfully completed.
-func (state PipelineRunState) SuccessfulPipelineTaskNames() []string {
-	done := []string{}
+// SuccessfulOrSkippedDAGTasks returns a list of the names of all of the PipelineTasks in state
+// which have successfully completed or skipped
+func (state PipelineRunState) SuccessfulOrSkippedDAGTasks(d *dag.Graph) []string {
+	tasks := []string{}
 	for _, t := range state {
-		if t.TaskRun != nil {
-			c := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-			if c.IsTrue() {
-				done = append(done, t.PipelineTask.Name)
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.IsSuccessful() || isSkipped(t, state.ToMap(), d) {
+				tasks = append(tasks, t.PipelineTask.Name)
+			}
+		}
+	}
+	return tasks
+}
+
+// isDAGTasksStopped returns true if any of the pipeline task has failed
+// and none of the pipleine task are still running
+func (state PipelineRunState) isDAGTasksStopped(d *dag.Graph) bool {
+	failed := false
+	for _, t := range state {
+		if t.IsFailure() {
+			failed = true
+			continue
+		}
+		if t.IsStarted() && !t.IsDone() {
+			failed = false
+			break
+		}
+	}
+	return failed
+}
+
+// isDAGTasksDone returns true if all DAG tasks are done executing,
+// DAG tasks could have failed/succeeded/skipped
+func (state PipelineRunState) isDAGTasksDone(d *dag.Graph) (isDone bool) {
+	isDone = true
+	for _, t := range state {
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.TaskRun == nil {
+				// this task might have skipped if taskRun is nil
+				// continue and ignore if this task was skipped
+				// skipped task is considered part of done
+				if isSkipped(t, state.ToMap(), d) {
+					continue
+				}
+				return false
+			}
+			isDone = isDone && t.IsDone()
+			if !isDone {
```

Comment:
What is the purpose of the variable `isDone` here?
it seems that this would achieve exactly the same:
```
if t.isDone() {
    return
}
```

Reply:
`isDone` is the expected return value which is defined in the function definition, this function is created based on [state.IsDone() ](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L153) function.

changing the `if` condition wouldn't achieve the same without updating named return value (`isDone`).

Let me know how you want to address this:

Option 1:

Change only this function to avoid using function definition, from

```
func (state PipelineRunState) isDAGTasksDone() (isDone bool) {
```

to

```
func (state PipelineRunState) isDAGTasksDone() bool {
```

Option 2: Leave as is and refactor this function and `state.IsDone()` to avoid using named return value

Option 3: Leave both functions (`isDAGTasksDone` and `state.IsDone()`) as is, they way it's implemented.

Reply:
Depending on what option we choose, I will create an issue to track potential refactoring efforts.

Reply:
Please feel free to add your pick to issue #2852

Reply:
Ah, thanks for the explanation, I didn't realise there was a named return.

NIT: Still I don't see the reason for the `&&`. The following:
```
isDone = isDone && t.IsDone()
```
Could simply be:
```
isDone = t.IsDone()
```

In the beginning `isDone` is initialised to `true` and ` && true` does nothing.
The moment the value turns of `isDone`, we return from the function.

It's not really important, only when I read the code I spend some time trying to figure out what that achieves.

Reply:
named return are so confusing, I would really like not having it (even if it means it makes the function more verbose)